core-lib/Streams.ns

```diff
@@ -0,0 +1,1037 @@
+				self position: origPos.
+				^cltn
+		)
+
+		public size ^ <Integer> = (
+			(* Make sure we flush our buffer if needed to ensure that
+			the buffer contents are taken into account *)
+
+			self flush.
+			^self descriptor size.
+		)
+
+		public updateBufferLocationAfterFlush = (
+			(* We don't advance the buffer here because we need to preserve the read semantics *)
+		)
+
+		writeIfNeededBeforePositionChange = (
+				self dirty
+						ifTrue: [ self writeBuffer; dirty: false. ].
+		)
+	) : (
+		public onDescriptor: d <DESCRIPTOR> ^ <Instance> = (
+				^self new initDescriptor: d
+		)
+	)
+
+	class ExternalStream = Object (
+	(* This is the abstract superclass for all external stream classes.  
+	It provides buffer management for subclasses.   
+
+	Invariant: the buffer position is always less than the index of the last buffer element.   
+	This is accomplished by proactively filling/flushing the buffer when the end is reached.*)
+  | 
+	descriptor <DESCRIPTOR> (* = d. *)
+	buffer <ExternalProxy> (* I/O buffer, cached from descriptor for speed (0 indexed) *)
+	bufferSize <Integer> (* I/O buffer size, cached from descriptor for speed *)
+	bufferOrigin <Integer> (* the offset between position and buffer coordinates *)
+	public nextBufferPosition <Integer> (* The actual position, in buffer coordinates, of the next element (i.e. position + 1 in buffer coords) *)
+	public bufferNumElements <Integer> (* The number of valid data bytes in the buffer *)
+	|) (
+		advanceBuffer = (
+				(* advance the buffer, and fill it if the receiver is readable.  Assumption: the last advanceBuffer completely
+					filled the buffer *)
+				self bufferOrigin: self bufferOrigin + self bufferSize.
+				self fillBuffer.
+				self nextBufferPosition: 0.
+		)
+
+		public close = (
+				self doClose
+		)
+
+		doClose = (
+				(* This is factored out from close so that subclasses that mixin BasicInputStream can reimplement close to call this *)
+				self descriptor close.
+				buffer:: nil.
+		)
+
+		fillBuffer = (
+				(* fill the buffer if needed (this default implementation works for noninput streams) *)
+			self bufferNumElements: 0
+		)
+
+		initBuffer = (
+			buffer:: descriptor buffer.
+			bufferSize:: descriptor bufferSize.
+		)
+
+		public initDescriptor: d <DESCRIPTOR>  = (
+			doInitDescriptor: d.
+		)
+
+		doInitDescriptor: d <DESCRIPTOR> = (
+			self descriptor: d.
+			self initBuffer.
+
+			self bufferOrigin: 0.
+			self bufferNumElements: 0.
+			self nextBufferPosition: 0.
+		)
+
+		postCopy = (
+			super postCopy.
+			self descriptor: self descriptor copy.
+			self initBuffer.
+		)
+	) : (
+		onDescriptor: d <DESCRIPTOR> ^ <Instance> = (
+				^self new initDescriptor: d
+		)
+	)
+
+	usingCrLfConvention = (
+		(* The current code only distinguishes systems that use CR and systems use CRLF for line endings. In order to generate files that are printed correctly on systems that use LF (Linux, OS X), we default to CRLF for now. *)
+		^ true
+	)
+) : (
```

Comment:
This is unfortunately really to much for me to review.
I don't have a good understanding of what's going on, and what the overall design is :(